Add support for cracking Monero wallets

kholia opened the pull request:
NOTE: I still have to clean up this code!

This fixes https://github.com/magnumripper/JohnTheRipper/issues/3144.

kholia commented:
Things to do,

* Reuse the existing Keccak code we have.

* Formatting, and whitespace cleanups.

kholia commented:
This is a very slow hash,

```
$ ../run/john --test --format=monero  # i7-6600U CPU
Will run 4 OpenMP threads
Benchmarking: monero, monero Wallet [AES+ChaCha+Various 64/64]... (4xOMP) DONE
Raw:	9.1 c/s real, 2.4 c/s virtual
```

magnumripper commented on src/monero_fmt_plug.c:
```diff
@@ -0,0 +1,270 @@
+/*
+ * Format for cracking Monero wallet hashes.
+ *
+ * This software is Copyright (c) 2017, Dhiru Kholia <dhiru [at] openwall.com>,
```
It's 2018!

magnumripper commented on src/monero_fmt_plug.c:
```diff
@@ -0,0 +1,270 @@
+#define FORMAT_NAME             "monero Wallet"
+#define FORMAT_TAG              "$monero$"
+#define TAG_LENGTH              (sizeof(FORMAT_TAG) - 1)
+#define ALGORITHM_NAME          "AES+ChaCha+Various 64/" ARCH_BITS_STR
```
We don't use to have `+` in algo names. I'd prefer spaces or perhaps `/`

magnumripper replied:
...and perhaps it should say "pseudo AES" or something, since it's not a normal one.

magnumripper commented on src/monero_fmt_plug.c:
```diff
@@ -0,0 +1,270 @@
+	return saved_key[index];
+}
+
+void cn_slow_hash(const void *data, size_t length, char *hash);
```
Perhaps create (and source) a `slow_hash.h`?

kholia commented:
I have addressed your code comments, thanks!

I am now testing cracking of Unicode passwords.

kholia commented:
Update,

* I have added a Unicode test vector now.

* Our BLAKE2 implementation was somehow not working. Added BLAKE2 implementation from Monero project.

I don't like the duplicated implementations. Hopefully, we will be able to unify them somehow.

kholia commented:
Monero wallet code base also supports `legacy` wallets. Support for them needs to be added and tested.

kholia commented:
Maybe we don't need to support the old wallet format. Even the oldest downloadable release of Monero (i.e. 0.9.1 Hydrogen Helix) uses the new JSON wallet format.

```
$ head monero2john.py
...
# + Tested with monero-gui-v0.11.1.0 on Fedora 27.
# + Tested with monero.linux.x64.v0-9-0-0.tar.bz2 (from Jan, 2016) on Fedora 27.
```

magnumripper commented:
> Mapping the new Keccak "API" to our existing Keccak implementation doesn't look trivial.

In what way was it not trivial? I just had a (very) quick look and I didn't see anything weird. Can't you just write wrapper macros?

> Our BLAKE2 implementation was somehow not working. Added BLAKE2 implementation from Monero project.

Do they produce different hashes? This should be investigated. I see there are flavours of Blake2... Here's test vectors for all of them: https://github.com/BLAKE2/BLAKE2/tree/master/testvectors

You have my "permission" to merge this as-is, although I'd rather see these issues fixed first (or it probably wont happen at all).

kholia commented:
I have opened an issue about this code duplication stuff, and assigned it to myself.
